core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala

```diff
@@ -116,22 +117,98 @@ class GroupCoordinator(val brokerId: Int,
     if (sessionTimeoutMs < groupConfig.groupMinSessionTimeoutMs ||
       sessionTimeoutMs > groupConfig.groupMaxSessionTimeoutMs) {
       responseCallback(joinError(memberId, Errors.INVALID_SESSION_TIMEOUT))
-    } else {
+      return
+    }
+
+    if (memberId == JoinGroupRequest.UNKNOWN_MEMBER_ID) {
+      doUnknownJoinGroup(groupId, requireKnownMemberId, clientId, clientHost, rebalanceTimeoutMs, sessionTimeoutMs, protocolType, protocols, responseCallback)
+      return
+    }
+
+    groupManager.getGroup(groupId) match {
+      case None =>
+        responseCallback(joinError(memberId, Errors.UNKNOWN_MEMBER_ID))
+      case Some(group) =>
+        doJoinGroup(group, memberId, clientId, clientHost, rebalanceTimeoutMs, sessionTimeoutMs, protocolType, protocols, responseCallback)
+    }
+  }
+
+  private def protocolNotMatch(group: GroupMetadata,
+                               memberId: String,
+                               protocolType: String,
+                               protocols: List[(String, Array[Byte])],
+                               responseCallback: JoinCallback): Boolean = {
+    val groupProtocolNotSupportedByNewMember = !group.is(Empty) &&
+      (!group.protocolType.contains(protocolType) || !group.supportsProtocols(protocols.map(_._1).toSet))
+    val firstMemberWithEmptyProtocolOrProtocolType = group.is(Empty) && (protocols.isEmpty || protocolType.isEmpty)
+    val notMatch = groupProtocolNotSupportedByNewMember || firstMemberWithEmptyProtocolOrProtocolType
+    if (notMatch) {
+      responseCallback(joinError(memberId, Errors.INCONSISTENT_GROUP_PROTOCOL))
```

Comment:
nit: it is a little surprising to have this method respond to the callback. I think it would be clearer to leave this as a simple check and let the caller send the response.

Reply:
done